lib/segment/src/id_tracker/immutable_id_tracker.rs

```diff
@@ -0,0 +1,582 @@
+            return true;
+        }
+        self.deleted[key]
+    }
+
+    fn name(&self) -> &'static str {
+        "immutable id tracker"
+    }
+
+    fn cleanup_versions(&mut self) -> OperationResult<()> {
+        let mut to_remove = Vec::new();
+        for internal_id in self.iter_internal() {
+            if self.internal_version(internal_id).is_none() {
+                if let Some(external_id) = self.external_id(internal_id) {
+                    to_remove.push(external_id);
+                } else {
+                    debug_assert!(false, "internal id {} has no external id", internal_id);
+                }
+            }
+        }
+        for external_id in to_remove {
+            self.drop(external_id)?;
+            #[cfg(debug_assertions)] // Only for dev builds
+            {
+                log::debug!("dropped version for point {} without version", external_id);
+            }
+        }
+        Ok(())
+    }
+
+    fn files(&self) -> Vec<PathBuf> {
+        vec![
+            Self::deleted_file_path(&self.path),
+            Self::mappings_file_path(&self.path),
+            Self::version_mapping_file_path(&self.path),
+        ]
+    }
+}
+
+impl<'de> Deserialize<'de> for PointMappings {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(PointMappingVisitor)
+    }
+}
+
+impl Serialize for PointMappings {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut sequence = serializer.serialize_seq(Some(self.internal_to_external.len()))?;
+
+        for external_id in self.internal_to_external.iter() {
+            let internal_id = match external_id {
+                StoredPointId::NumId(n) => self.external_to_internal_num.get(n),
+                StoredPointId::Uuid(u) => self.external_to_internal_uuid.get(u),
+                StoredPointId::String(str) => {
+                    unimplemented!("cannot convert internal string id '{str}' to external id")
+                }
+            }
+            .unwrap();
+
+            sequence.serialize_element(&(external_id, internal_id))?;
+        }
+
+        sequence.end()
+    }
+}
+
+mod point_mappings_deser {
+    use std::collections::BTreeMap;
+    use std::fmt::Formatter;
+
+    use common::types::PointOffsetType;
+    use serde::de::{SeqAccess, Visitor};
+    use uuid::Uuid;
+
+    use crate::id_tracker::immutable_id_tracker::PointMappings;
+    use crate::id_tracker::simple_id_tracker::StoredPointId;
+
+    pub(super) struct PointMappingVisitor;
+
+    impl<'de> Visitor<'de> for PointMappingVisitor {
+        type Value = PointMappings;
+
+        fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+            formatter.write_str("a sequence of tuples")
+        }
+
+        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
```

Comment:
```suggestion
        // Custom deserializer to prevent allocating a humongous intermediate list of points
        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
```